include/mitsuba/render/mesh.h

```diff
@@ -286,16 +241,41 @@ class MTS_EXPORT_RENDER Mesh : public Shape<Float, Spectrum> {
 
 #if defined(MTS_ENABLE_OPTIX)
     /// Return the OptiX version of this shape
-    virtual RTgeometrytriangles optix_geometry(RTcontext context) override;
+    virtual void optix_geometry() override;
+    virtual void optix_build_input(OptixBuildInput&) const override;
+    virtual void optix_hit_group_data(HitGroupData&) const override;
+
 #endif
 
     /// @}
     // =========================================================================
 
-
     /// Return a human-readable string representation of the shape contents.
     virtual std::string to_string() const override;
 
+private:
+
+    template<uint32_t Size>
+    auto eval_vertex_attribute_impl(const DynamicBuffer<Float> &buf, const SurfaceInteraction3f& si, Mask active = true) const {
+        using Result = Vector<replace_scalar_t<Mask, float>, Size>;
+
+        auto fi = face_indices(si.prim_index, active);
+        auto[b0, b1, b2] = barycentric_coordinates(si, active);
+
+        Result attrs[3] = { gather<Result>(buf, fi[0], active),
+                            gather<Result>(buf, fi[1], active),
+                            gather<Result>(buf, fi[2], active) };
+        Result attr = b0 * attrs[0] + b1 * attrs[1] + b2 * attrs[2];
+
+        return attr;
+    }
+
+    template<uint32_t Size>
+    auto eval_face_attribute_impl(const DynamicBuffer<Float> &buf, const SurfaceInteraction3f& si, Mask active = true) const {
+        using Result = Vector<replace_scalar_t<Mask, float>, Size>;
+        return gather<Result>(buf, si.prim_index, active);
+    }
+
```

Comment:
This could be combined in a single function to which we would pass an extra `MeshAttribute` arguement so it can figure out what interpolation/attribute type it is.